### PR TITLE
Report number of paramters trained by trainer.

### LINKED ISSUE #257: Report total number of network parameters (while building the model)
It would be great to report the total number of network parameters.

This could be even reported along with the gradual building of the model, ie. as part of issue #93, e.g. after every object being instantiated.

The total number of parameters is also asked for in the more general issue #109.

### PR DESCRIPTION
**Pull request by jlibovicky**
This PR introduces computing  of total number of parameters being optimized in a trainer. There are reported in the log at the beginning of the training loop.

It resolves #257.

### CODE REVIEW

File: neuralmonkey/trainers/generic_trainer.py [hunk 1]
```diff
@@ -25,8 +26,9 @@
 class GenericTrainer(object):
 
     def __init__(self, objectives: List[Objective],
-                 l1_weight=0.0, l2_weight=0.0,
-                 clip_norm=False, optimizer=None, global_step=None) -> None:
+                 l1_weight: float=0.0, l2_weight: float=0.0,
+                 clip_norm: Optional[float]=None, optimizer=None,
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: to tam bylo clip norm je bool a prošlo to jo? :-) to proto že to je všude nastavený na 1.0?

File: neuralmonkey/trainers/generic_trainer.py [hunk 2]
```diff
@@ -74,6 +76,8 @@ def __init__(self, objectives: List[Objective],
                              for grad, var in gradients
                              if grad is not None]
 
+            self.n_parameters = _number_of_params(gradients)
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: jak to tady je s těma parametrama toho Adama?

File: neuralmonkey/trainers/generic_trainer.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/trainers/generic_trainer.py]: asi jo

File: neuralmonkey/trainers/generic_trainer.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/trainers/generic_trainer.py]: Jsou to všechny trainable parametry. Ty svoje paramtery si Adam nějak počítá sám, ale neřekl bych o nich, že se trénujou. Podobně by to bylo třeba s batch normalizací - tam je zase rozptyl a střední hodnota, co se počítají, ale netrénují se. Za parametr modelu bych považoval něco, co se přímo trénuje.

Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: Já taky. Spíš mi jde o to, co tam Tensorflow láduje.

**Comment by obo** [comment]
Jeste nez to mergnete, tak bych rad pridal GPU meminfo do te radky. Tak prosim pockejte.

**Comment by jindrahelcl** [comment]
co to děláš, @obo? co je tedy #242?

**Comment by obo** [comment]
Tady je to (myslim) prvni uzitecne misto. #424 to dela u radky Epoch, tj. pravidelne pri behu. Tohle je hned po inicializaci a kdyz clovek nastavi epochs=0, ocekavam, ze uz uvidim, kolik pameti model sezral.

A epochs=0 je reseni #111 alias #267

**Comment by jindrahelcl** [comment]
takže řikáš, ať vyhodíme vypisování, jak zavedlo #242... ?

Příště si prosím udělej svuj pull request..

**Comment by obo** [comment]
Nechal bych urcite oboji reportovani, viz moje argumentace na #242 . Za mne bud podle spolecneho prepinace report_gpu_memory_usage (jak je to takto), nebo ten prepinac pouzivat jen pro to pravidelne reportovani a zde to zavolat vzdy.
Sorry, snazil jsem se omezit pocet pull requestu. Podprahove mi porad casto pripadaji zbytecny uredni overhead.

File: neuralmonkey/learning_utils.py [hunk 3]
```diff
@@ -65,6 +66,14 @@ def training_loop(tf_manager: TensorFlowManager,
...
     """
 
+    if tf_manager.report_gpu_memory_consumption:
+        meminfostr = ", GPU memory usage: "+gpu_memusage()
+    else:
+        meminfostr = ""
+
+    log("Model has {} trainable parameters{}."
+        .format(trainer.n_parameters, meminfostr))
+
```
Comment (jindrahelcl) [inline neuralmonkey/learning_utils.py]: hezčí by bylo
```
paramstr = "Model has {} trainable parameters".format(trainer.n_parameters)
if tf_manager.report_gpu_memory_consumption:
    paramstr += ", GPU memory usage: {}".format(gpu_memusage())

log(paramstr)
```

**Comment by jindrahelcl** [comment]
Já jsem původně pochopil "první užitečný místo" jako že to, jak to bylo doposud je naprd, ale asi si tim myslel, že tohle je nejdřív.

Tahle byrokracie je alespoň dobrá k tomu, že máme přehled o tom, co kdo udělal a co kde je, a co se kde mohlo pokazit, a čas od času se skutečně najde nějakej potenciální bug.. Nenápadný přišoupávání komitů jde spíš proti tomu.. :-)

Největší riziko by bylo v situaci, kdy by @jlibovicky nevěděl, žes tam něco přikomitoval, zjistil, že je pull request approvlej a udělal u sebe rebase na master a s `--force` to pushnul na github (což je momentálně standardní workflow) - a tvůj komit by se tim smazal.
